Add the ability to share resources between users.

thiagosil opened the pull request:
I tried to split my work into meaningful commits to make it easier to review.

### 1. Proposed solution
I created a polymorphic association to represent the ability to share resources between users.
This way it would be extensible to other models in the future as needed. 

![image](https://github.com/user-attachments/assets/cb48ae80-64c3-424c-86b7-85e4824d3890)

### 2. API
Based on the frontend I created 3 routes that will be required:
 - Create a Shared Resource
  ```
  POST /shared_resources

body
{ 
shared_resource: { shareable_id: project.id, shareable_type: 'Project', user_id: user.id, permission_level: 'full_access' }
}
```` 
 - Update a shared resource permission level
 ```
 PUT shared_resources/:resource_type/:resource_id/:user_id
body 
{
shared_resource: { permission_level: 'view}
}
```
 - Show Shared Resource
` GET /shared_resources/:id`

A list shared resources route will also be required in the future to list existing shared resources for a project, document, and report.
`GET /shared_resources/project/:project_id`

### 3. Issues Found 
I found a few issues when testing the app:
1 - The user id was not being returned so react was displaying a duplicate key error.
```
    {users?.map?.((user: User) => (
        <div key={user.id} className="flex items-center justify-between bg-gray-700 p-2 rounded">
```
2 - When a new user is created, the name is not currently required, AvatarFallback was also displaying an error.
  ```
  <Avatar>
        <AvatarImage src={user.avatar_url} alt={user.name} />
        <AvatarFallback>{user.name?.charAt(0)}</AvatarFallback>
    </Avatar>
```
3 - Devise views are showing an error, this one I did not fix, this due time constraints
```
Uncaught Error: createRoot(...): Target container is not a DOM element.
    createRoot http://localhost:5000/assets/application-9492d25cd2224e22fcc40e2e2c8776209622d4d6cc9bb478b68a50d103b9fc13.js:23104
```


### 4. Comments about the project itself
I ended up using a little bit more than 2 hours(about 2.5 hours). I tried to cover the backend mostly, and to add tests of scenarios that I saw fit.
I was able to create a model, and the API, and started listing shared user projects using the projects index API.
I manually created some shared projects to test using the frontend, and for the success scenario it worked well.

### 5. Next Steps

For something bigger than the scope of a take-home test I would not try to reinvent the wheel and use something like can can can.
Another thing I would do is do a proper authorization verification before executing any user actions.

1 - Replicate share Projects behavior to Documents and Reports.
2 - Implement API hooks on the frontend, then add actual functionality.
3 - Authorize user actions based on user permissions.
4 - Use authorization gem like pundit or can can can.

That was a fun project to do. It gave me a lot to reflect on the best way of implementing the solution to future-proof the implementation and how to choose what should be prioritized. In case you want to discuss anything further just let me know.

I used copilot to generate some of the tests, mainly for boilerplate code.

tez123z commented on config/routes.rb:
```diff
@@ -1,14 +1,18 @@
   resources :users, only: [:index]
+  resources :shared_resources, only: [:show, :create]
+
+  put 'shared_resources/:resource_type/:resource_id/:user_id', to: 'shared_resources#update_permission'
```
This approach is interesting - any reason we couldn't just use a more generic update endpoint and only allow certain parameters? Is there a particular reason you chose this pattern instead?

thiagosil replied:
The only reason I did that was because I wanted to be more explicit about what this API does.
But I don't have a strong argument for or against any of those endpoints, I would be open to either. This would probably depend on who would be using this API. What would be the scenarios we would like to cover with Shared Resources.

tez123z commented on app/views/users/_user.json.jbuilder:
```diff
@@ -1 +1 @@
-json.extract! user, :name, :avatar_url, :email, :created_at, :updated_at
+json.extract! user, :id, :name, :avatar_url, :email, :created_at, :updated_at
```
👏 Good find!

gastonsoto commented on app/models/user.rb:
```diff
@@ -8,5 +8,13 @@ class User < ApplicationRecord
+  has_many :shared_reports, through: :shared_resources, source: :shareable, source_type: 'Report'
+  has_many :shared_documents, through: :shared_resources, source: :shareable, source_type: 'Document'
+
+  def user_projects
+    created_projects + shared_projects
+  end
```
Adding `created_projects` and `shared_projects` like this, affects pagination. What would be your approach to support pagination?

thiagosil replied:
To support pagination I would add scopes to the project model,  this way we would have a single query for user_projects.

Something like this:

``` 
 scope :shared_with, ->(user) {
    where(id: SharedResource.where(
      shareable_type: 'Project',
      user_id: user.id
    ).select(:shareable_id))
  }


  scope :accessible_by, ->(user) {
    created_by(user).or(shared_with(user))
  }

thiagosil replied:
To support pagination I would add scopes to the project model, this way we would have a single query for user_projects.

Project Scopes:

```
 scope :shared_with, ->(user) {
    where(id: SharedResource.where(
      shareable_type: 'Project',
      user_id: user.id
    ).select(:shareable_id))
  }


  scope :accessible_by, ->(user) {
    created_by
(user).or(shared_with(user))
  }
```  
  
  And change user projects to something like this:
  
  ```
    def user_projects
      Projects.accessible_by(self)
    end

thiagosil replied:
I did an initial version of that. You can check the last commits.

tez123z commented on spec/models/user_spec.rb:
```diff
@@ -8,4 +8,38 @@
+
+        context 'when user has shared projects' do
+            let(:shared_project) { create(:project) }
+            let(:shared_resource) { create(:shared_resource, shareable: shared_project, user: user) }
+
+            before { shared_resource }
```
```suggestion
            let!(:shared_resource) { create(:shared_resource, shareable: shared_project, user: user) }
```

`let!` will init the factory before the test here.

thiagosil replied:
Thanks for catching that, fixed.

gastonsoto commented on config/routes.rb:
```diff
@@ -1,14 +1,18 @@
   resources :users, only: [:index]
+  resources :shared_resources, only: [:show, :create]
+
+  put 'shared_resources/:resource_type/:resource_id/:user_id', to: 'shared_resources#update_permission'
```
I have the feeling we can make it more extensible by using nested resources. What do you think? What would be your approach here for a second iteration?

thiagosil replied:
I would probably discuss what we want to accomplish with this APIs, what features we want to add. But that certainly could be a good thing to do. Since we will want to add shared resources for different kinds of models.

gastonsoto commented on config/routes.rb:
```diff
@@ -1,14 +1,18 @@
 Rails.application.routes.draw do
+  get 'shared_resources/create'
```
NIT: Having `shared_resources` below, what are your thoughts here with this `GET` route pointing to a `create` method?

thiagosil replied:
Thanks for catching that. I used rails generator that created this route. I just removed it. The create route should be a POST.

thiagosil commented:
@gastonsoto @tez123z I added a few commits with further improvements. Also I changed the update_permission API to create the permission when it doesn't exist to enable you to test using the frontend.

gastonsoto commented:
> @gastonsoto @tez123z I added a few commits with further improvements. Also I changed the update_permission API to create the permission when it doesn't exist to enable you to test using the frontend.

Thanks @thiagosil ! Great work!

tez123z commented:
Great work @thiagosil!
